Add improved functionality for the construction of json_arrayt and json_objectt

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->
Much of the existing code which constructs instances of `json_arrayt` and `json_objectt` are based around creating an empty instance as then mutating it to add the contained information. This PR adds extra functionality to enable these two types to be constructed using a coding style which no longer requires mutation and is more readable.

thomasspriggs committed:
Add initializer_list constructor to `json_objectt`

This new constructor facilitates the construction of instances of
`json_objectt` which are const, because it gives a tidy way to construct
an entire `json_objectt`.

thomasspriggs committed:
Add initializer_list constructor to `json_arrayt`

This new constructor facilitates the construction of instances of
`json_arrayt` which are const, because it gives a tidy way to construct
an entire `json_arrayt`.

thomasspriggs committed:
Add support for range to construct json_arrayt and json_objectt

The end result of this commit is that code like the following example
can be used to construct `json_arrayt` / `json_objectt` -
```
const std::vector<std::string> input{"foo", "bar"};
const auto json_array =
make_range(input)
.map(constructor_of<json_stringt>())
.collect<json_arrayt>();
```
This commit includes -
* Constructors from iterators for json_arrayt and json_objectt, so
that the iterators from a range can be used to construct these classes.
* A `collect` member function for `ranget`, so that a chain of range
operations can finish with the construction of a resulting container
containing a collection of the results.
* A `constructor_of` template function, which provides syntactic sugar
when using `map` to call a constructor, compared to writing a new lambda
function each time such an operation is carried out.
* Unit tests covering all of the above functionality.

thomasspriggs committed:
Refactor `static_verifier_json` using `range`

This commit is included in this PR, as a demonstration of the coding
style which is enabled by the proceeding commits.

thomasspriggs committed:
Add cast operator from `ranget` to a container

Adding `.collect` to the end of a chain of range operations can add
clutter to the code, rather than improving readability, in places where
the type of container being constructed is apparent from the context.
This commit adds a cast operation in order to improve readability in
this scenario.

In addition to the operator, this commit includes a refactor of
`optionst::to_json` in order to demonstrate an example usage of the
proposed change.

tautschnig commented on src/util/json.h:
```diff
@@ -283,6 +283,12 @@ class json_objectt:public jsont
 
+  explicit json_objectt(
+    std::initializer_list<typename objectt::value_type> initializer_list)
+    : json_objectt{objectt{initializer_list}}
```
Thank you for this further improvement! I suspect it's kind of a painful ask, but would you mind adding a commit that now removes the outermost parentheses from various bits touched in 64df79edce7dec4f?

tautschnig commented on src/goto-analyzer/static_verifier.cpp:
```diff
@@ -16,52 +16,47 @@ Author: Martin Brain, martin.brain@cs.ox.ac.uk
 #include <goto-programs/goto_model.h>
 
 #include <analyses/ai.h>
+#include <util/range.h>
```
Please move this 5 lines up so that it becomes part of the `util` block.

thomasspriggs committed:
Remove unnecessary brackets in constructon of `json_objectt`

With the new constructors, these brackets are no longer required and can
be removed, in order to improve readability.

thomasspriggs committed:
Remove unused constructors from backing container

Removing this constructor simplifies the interface, by reducing the
number of different ways to construct the classes and avoiding exposing
the implementation detail of what type of container `json_objectt` and
`json_arrayt` are backed by.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 55051eabf7446bd61de69f07ea6f9c23e924fd9d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97979428

tautschnig reviewed (approved):
Thank you for the extra work!

kroening commented:
Does this work with Visual Studio 2013?

thomasspriggs commented:
@kroening I wrote this PR under Debian 9 "Stretch" using gcc 6. So I haven't tried building in VS 2013 myself. However the windows CI build has passed.

thk123 commented on src/util/range.h:
```diff
@@ -351,6 +351,14 @@ struct ranget final
+  /// Constructs a collection containing the values, which this range iterates
+  /// over.
+  template <typename containert>
+  containert collect() const
```
:green_heart:

thk123 commented on src/util/constructor_of.h:
```diff
@@ -0,0 +1,33 @@
+
+/// A type of functor which wraps around the set of constructors of a type.
+/// \tparam constructedt: The type which this functor constructs.
+template <typename constructedt>
```
:question: What is the use of this? Currently only used in the unit test so was thinking it should live there, but of course if it has general utility then here is fine

thk123 replied:
Is used in a later commit.

thk123 commented on src/util/range.h:
```diff
@@ -359,6 +359,12 @@ struct ranget final
     return containert(begin(), end());
   }
 
+  template <typename containert>
```
:green_heart: :green_heart:

thk123 reviewed (approved):
Love it.

thomasspriggs commented:
@kroening Do you want any additional checks done around visual studio 2013 compatibility? CI passing + reviewer approval would normally be considered to be sufficient. Given that the new code is both unit tested and used, there should have been compile errors, or failing tests if any issues were introduced.

This PR now touches a wide range of directories, due to the bracket removal requested by Michael Tautschnig. Which means that it seems to require review from a unknown number of reviewers. It would save time if you would merge this.

smowton commented on src/goto-analyzer/static_verifier.cpp:
```diff
@@ -20,48 +21,42 @@ Author: Martin Brain, martin.brain@cs.ox.ac.uk
 };
 
+/// Makes a status message string from a status.
+static std::string message(const static_verifier_resultt::statust &status)
```
Suggest `const char *` since this only ever returns constants (no need to bind our caller to necessarily go through `std::string`)

thomasspriggs replied:
Done.

smowton commented on src/goto-analyzer/unreachable_instructions.cpp:
```diff
@@ -152,8 +152,8 @@ static void add_to_json(
     const source_locationt &l=it->second->source_location;
-    json_objectt i_entry(
-      {{"sourceLocation", json(l)}, {"statement", json_stringt(s)}});
+    json_objectt i_entry{{"sourceLocation", json(l)},
```
Surprising formatting -- break after first `{`?

thomasspriggs replied:
If I insert a new line, then clang format will just remove it again. I agree that left alignment of the enclosed pairs would be preferable. However I prefer the sub optimal formatting over the extra comments to switch off the formatting in this case.

smowton commented on src/goto-programs/json_expr.cpp:
```diff
@@ -172,8 +172,8 @@ json_objectt json(const typet &type, const namespacet &ns, const irep_idt &mode)
     {
-      json_objectt e({{"name", json_stringt(component.get_name())},
-                      {"type", json(component.type(), ns, mode)}});
+      json_objectt e{{"name", json_stringt(component.get_name())},
```
Same formatting nitpick as above

smowton commented on src/util/json.h:
```diff
@@ -160,8 +161,24 @@ class json_arrayt:public jsont
+  {
+  }
+
+  template <typename begin_iteratort, typename end_iteratort>
```
Surely these have to be the same type?

thomasspriggs replied:
Forwarding these parameters separately means that the template can match when the two iterators have different `const` or reference qualifiers. For example `.begin()` on a range returns a mutable instance of the iterator by value, whereas the `.end()` on a range returns a const reference to the end iterator.
I will however add a `static_assert` to these constructors, in order to give a sensible error message if the types of the iterators are different once any `const` or reference modifiers are removed.

smowton commented on src/util/options.cpp:
```diff
@@ -90,14 +92,12 @@ const optionst::value_listt &optionst::get_list_option(
 /// Returns the options as JSON key value pairs
```
Worth noting this is now lazy, so calling this then mutating `optionst` then trying to iterate over the range may be undefined behaviour

thomasspriggs replied:
Please note that the return type of the function is still a value of type `json_objectt`, not a range. So there is no laziness beyond the scope of this function.

hannes-steffenhagen-diffblue replied:
@thomasspriggs might be worth adding an explicit constructor call to make that more obvious?

smowton reviewed (approved):
(notes above, meant to approve)

kroening commented:
@thomasspriggs The CI uses 2015, so may still fail with 2013.
I can try tomorrow.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: a8b278c7e1be7d5ed687c10b21caf545804fe9f6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98067771
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a859b172de4bffd47eceacc4fb60958e8b026f1e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98072678

thomasspriggs commented:
@kroening I have made a cursory attempt at compiling cbmc **`develop`** with visual studio 2013. This resulted in numerous build errors. Some of these were likely to be due to not having my build environment correctly setup. However a substantial number of these appear to be existing compatibility issues.

kroening commented:
This doesn't compile with VS 2013.
I.e., merging this requires ditching this target.

thomasspriggs commented:
Can you compile `develop` with VS 2013? If you can't, then VS 2013 support has already been lost without this PR.

thomasspriggs commented:
I have pushed two rebases to 
 * Rename `constructort` to `constructor_oft` to avoid naming conflicts.
 * Rebase on develop. Resolving a merge conflict in the process.

thomasspriggs commented:
Rebased to fix include of `catch` in unit test.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 2a4d8ff49767b76b3bc032ab2d93de66b05738c0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98867160
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

martin-cs reviewed (approved):
I don't see a problem with my part.
